### ISSUE TITLE
XGBoost in Spark "stops" for a couple hours, then continues

### ISSUE THREAD

**Issue by olbapjose** [dmlc/xgboost#3470]
Spark version 2.2.0. 
XGBoost version: 0.80-SNAPSHOT
Cluster with CentOS, 3 workers, 90 GB RAM.

Apparently, when training under Spark (Scala), I can see the Spark logs when each tree is fit, something like: 

[14:20:03] /home/myuser/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 38122 extra nodes, 0 pruned nodes, max_depth=15

I can see a new tree after every few seconds but at some point, sometimes after one hour or two, I see no new messages when refreshing the log. Then after two or three hours "in pause", the process continues and I can see new trees again added to the log. This does not always happen but it does happen quite often.

Is there any explanation for this? 

Thanks

**EDIT:** I thought it could be related to memory space but it seems it is not as there's much free memory right now:

![image](https://user-images.githubusercontent.com/34943494/42734354-cdfa8eee-8842-11e8-8528-055914d3d40b.png)

**Comment by RAMitchell** [comment]
> 38122 extra nodes

This is a pretty huge decision tree are you sure that is what you want?

**Comment by CodingCat** [comment]
most of memory usage in training set is in native memory space, so it will not be displayed in Spark UI

if you are using YARN and you are sure that your on-heap space is lightly used, you can try to set spark.executor.memoryOverhead larger to allocate more for XGBoost

**Comment by pvillacorta-stratio** [comment]
Hi all,

Thank for your answers. I am using Mesos actually. Following your suggestions, I just moved to a different configuration with just max depth 5 and a larger number of trees. Apparently it is now working well in one of the clusters, but I am getting an error in another cluster just at the beginning of the training, 
`java.lang.InterruptedException: ExecutorLost during XGBoost Training: ExecutorLostFailure (executor 0 exited caused by one of the running tasks) Reason: Remote RPC client disassociated`
which is quite misterious but I guess is caused by some issue of bad configuration of number or workers or something like that.

**Comment by eshengit** [comment]
could you take a thread dump for the container node which having active tasks to see whats going on?

**Comment by CodingCat** [comment]
any update on this?

**Comment by worldwar2008** [comment]
any update on this? I have the same problem when I followed the tutorial https://xgboost.readthedocs.io/en/latest/jvm/xgboost4j_spark_tutorial.html

**Comment by olbapjose** [comment]
As I said, I fixed it by fitting 5-depth trees.

Later I also discovered that, when Mesos does not give my job the exact number of executors I want (despite setting SPARK_CORES_MAX and SPARK_EXECUTOR_CORES explicitly), the training part never starts because in those situations I get less executors than the number I passed as the N_WORKERS option. If both do not match, it remains running but doing nothing without showing a warning.

Moreover, the InterruptedException was caused (I think, it was some time ago already) by not setting the locale env variables to "en_US" in all the workers, which is something I have suffered from the beginning. Since I am using Spark Job Dispatcher in a production env, it was a bit tricky until I found out how to set them. I had already solved it in my devel environment without Mesos and Job Dispatcher, just plain spark-submit.

**Comment by pvillacorta-stratio** [comment]
(Sorry for my double identity haha)

As I said, I fixed it by fitting 5-depth trees.

Later I also discovered that, when Mesos does not give my job the exact number of executors I want (despite setting SPARK_CORES_MAX and SPARK_EXECUTOR_CORES explicitly), the training part never starts because in those situations I get less executors than the number I passed as the N_WORKERS option. If both do not match, it remains running but doing nothing without showing a warning.

Moreover, the InterruptedException was caused (I think, it was some time ago already) by not setting the locale env variables to "en_US" in all the workers, which is something I have suffered from the beginning. Since I am using Spark Job Dispatcher in a production env, it was a bit tricky until I found out how to set them. I had already solved it in my devel environment without Mesos and Job Dispatcher, just plain spark-submit

**Comment by thvasilo** [comment]
I think the behavior @pvillacorta-stratio  mentioned is consistent in a YARN cluster: If a user requests more containers than are available the job will stall. I'm not sure if that should be considered a bug, since it's up to the YARN scheduler to find more containers, but it's sometimes unexpected behavior from the user perspective: They try to launch a job with X workers on a cluster that has less than X workers (common when first trying out distributed training on a test cluster) and the job never starts.

Ideally we'd like to have message of some kind to inform that not enough workers are available, but that is probably on the dmlc-core side.

**Comment by CodingCat** [comment]
please read https://xgboost.readthedocs.io/en/latest/jvm/xgboost4j_spark_tutorial.html#gang-scheduling

**Comment by pvillacorta-stratio** [comment]
Thank you all for your answers. It was Mesos, not YARN, but yes that's the point. The job starts but there is a foreachPartition in which all tasks complete successfully except for the last K tasks, where K is the number of workers. It is in those last K tasks where the trees are fit to the data. Those K tasks never complete (never start) but no error is given.
